src/proxy-animation.js

```diff
@@ -29,152 +29,518 @@ function createReadyPromise(details) {
+      // functions with closer adherence to the spec.
+
+      // Apply pending playbackRate
+      if (details.pendingPlaybackRate !== null) {
+        const previousCurrentTime = details.animation.currentTime;
+        details.animation.playbackRate = details.pendingPlaybackRate;
+        const timelineTime = details.timeline.currentTime;
+        details.startTime = details.pendingPlaybackRate ?
+            timelineTime - previousCurrentTime / details.pendingPlaybackRate :
+            previousCurrentTime;
+        details.pendingPlaybackRate = null;
+      }
+      switch(details.playState) {
+        case 'paused':
+          details.startTime = null;
+          break;
+
+        case 'running':
+        case 'finished':
+          details.holdTime = null;
+      }
+      resolve();
     });
   });
 }
 
+function effectivePlaybackRate(details) {
+  if (details.pendingPlaybackRate)
+    return details.pendingPlaybackRate;
+  return details.animation.playbackRate;
+}
+
+function applyPendingPlaybackRate(details) {
+  if (details.pendingPlaybackRate) {
+    details.animation.playbackRate = details.pendingPlaybackRate;
+    details.pendingPlaybackRate = null;
+  }
+}
+
+function updateFinishedState(details) {
+  if (!details.timeline)
+    return;
+
+  // No change in play state if idle or paused.
+  if (details.playState != 'running' && details.playState != 'finished')
+    return;
+
+  const isFinished = () => {
+    const playbackRate = effectivePlaybackRate(details);
+    const currentTime = details.animation.currentTime;
+    if (playbackRate < 0 && currentTime <= 0)
+      return true;
+    if (playbackRate > 0 &&
+        currentTime >= effectEnd(details))
+      return true;
+
+    return false;
+  };
+
+  const newPlayState = isFinished() ? 'finished' : 'running';
+  if (newPlayState == details.playState)
+    return;
+
+  details.playState = newPlayState;
+  if (newPlayState == 'finished') {
+    requestAnimationFrame(() => {
+      // Finished state may have been temporary. Ensure that we are still in the
+      // 'finished' state.
+      if (details.playState == 'finished') {
+        // Resolve the finished promise and queue the onfinished event.
+        // Finish snaps to the boundary. Restore current time after the finish
+        // call.
+        // TODO: Polyfill the finished promise.  If the timeline goes inactive
+        // during the lifetime of the animation then the finished promise for
+        // the underlying animation will be rejected.
+        const previousCurrentTime = details.animation.currentTime;
+        details.animation.finish();
+        details.animation.pause();
+        details.animation.currentTime = previousCurrentTime;
+      }
+    });
+  }
+}
+
+function effectEnd(details) {
+  const timing = details.animation.effect.getTiming();
+  const totalDuration =
+     timing.delay + timing.endDelay + timing.iterations * timing.duration;
```

Comment:
FYI activeDuration (used by calculateTargetEffectEnd) already accounts for iterations * duration.